nodejs/eks/nodegroup.ts

```diff
@@ -677,8 +695,50 @@ export type ManagedNodeGroupOptions = Omit<aws.eks.NodeGroupArgs, "clusterName"
     scalingConfig?: pulumi.Input<awsInputs.eks.NodeGroupScalingConfig>
 };
 
-export function createManagedNodeGroup(name: string, args: ManagedNodeGroupOptions, parent: pulumi.ComponentResource): aws.eks.NodeGroup {
+export function createManagedNodeGroup(name: string, args: ManagedNodeGroupOptions): aws.eks.NodeGroup {
     const core = isCoreData(args.cluster) ? args.cluster : args.cluster.core;
+    const parent = isCoreData(args.cluster) ? args.cluster.cluster : args.cluster;
+
+    if (!args.nodeRole && !args.nodeRoleArn) {
+        throw new Error(`An IAM role, or role ARN must be provided to create a managed node group`);
+    }
+
+    if (args.nodeRole && args.nodeRoleArn) {
+        throw new Error("nodeRole and nodeRoleArn are mutually exclusive to create a managed node group.");
+    }
+
+    // Compute the nodegroup role.
+    let roleArn: pulumi.Input<string> = pulumi.output("");
```

Comment:
This `pulumi.output("")` will never be used right?  Odd to have it here if there is an invariant that `roleArn` will always have a real value.

Reply:
Right, it's only for initialization.

FWIW we have used the same pattern in the EKS repo, e.g.

- `clusterSubnetIds`:
https://github.com/pulumi/pulumi-eks/blob/0579bf3c8ac5f611b55698b8bc8d24c5096b51b3/nodejs/eks/cluster.ts#L274-L296

If we drop the `= pulumi.output("")`, its use below in `nodegroupRole` is not compilable as tsserver states the `roleArn` is used before being assigned:

![image](https://user-images.githubusercontent.com/1112768/73481295-62aa6400-4350-11ea-8c19-209140f88351.png)

What do you suggest we do instead?

Reply:
I'm guessing this is because the following if/else statement doesn't cover all cases. You could rewrite it to include the error cases, and that should fix it.

Reply:
Resolved